File: frontend/src/lib/api/programmi.ts
/**
 * Wrapper API per `/api/programmi/*` (Sprint 4.3 backend).
 *
 * Tipi allineati a `colazione.schemas.programmi`:
 * - `ProgrammaMaterialeRead` / `ProgrammaDettaglioRead`
 * - `ProgrammaMaterialeCreate` / `ProgrammaMaterialeUpdate`
 * - `ProgrammaRegolaAssegnazioneRead`
 *
 * Le date sono ISO `YYYY-MM-DD`; i timestamp `created_at`/`updated_at`
 * sono ISO 8601 con timezone.
 */

import { apiJson } from "@/lib/api/client";

export type ProgrammaStato = "bozza" | "attivo" | "archiviato";

/**
 * Sprint 8.0 MR 0 (entry 164): pipeline state machine ramo PdC.
 * Mantenere allineato con `colazione.domain.pipeline.StatoPipelinePdc`.
 */
export type StatoPipelinePdc =
  | "PDE_IN_LAVORAZIONE"
  | "PDE_CONSOLIDATO"
  | "MATERIALE_GENERATO"
  | "MATERIALE_CONFERMATO"
  | "PDC_GENERATO"
  | "PDC_CONFERMATO"
  | "PERSONALE_ASSEGNATO"
  | "VISTA_PUBBLICATA";

/**
 * Sprint 8.0 MR 0 (entry 164): pipeline state machine ramo Manutenzione.
 * Mantenere allineato con `colazione.domain.pipeline.StatoManutenzione`.
 */
export type StatoManutenzione =
  | "IN_ATTESA"
  | "IN_LAVORAZIONE"
  | "MATRICOLE_ASSEGNATE";

const STATI_PIPELINE_PDC_ORDER: ReadonlyArray<StatoPipelinePdc> = [
  "PDE_IN_LAVORAZIONE",
  "PDE_CONSOLIDATO",
  "MATERIALE_GENERATO",
  "MATERIALE_CONFERMATO",
  "PDC_GENERATO",
  "PDC_CONFERMATO",
  "PERSONALE_ASSEGNATO",
  "VISTA_PUBBLICATA",
];

/**
 * Sprint 8.0 MR 1 (entry 165): replica della logica `materiale_freezato`
 * server-side. Quando `True`, il pianificatore non può modificare
 * regole/parametri/giri (PATCH/POST regole/genera-giri ritornano 409).
 */
export function materialeFreezato(stato: StatoPipelinePdc): boolean {
  const idx = STATI_PIPELINE_PDC_ORDER.indexOf(stato);
  const soglia = STATI_PIPELINE_PDC_ORDER.indexOf("MATERIALE_CONFERMATO");
  return idx >= 0 && idx >= soglia;
}

export interface StrictOptions {
  no_corse_residue: boolean;
  no_overcapacity: boolean;
  no_aggancio_non_validato: boolean;
  no_orphan_blocks: boolean;
  no_giro_appeso: boolean;
  no_km_eccesso: boolean;
}

export interface ProgrammaMaterialeRead {
  id: number;
  azienda_id: number;
  nome: string;
  valido_da: string;
  valido_a: string;
  stato: ProgrammaStato;
  /** Sprint 8.0 MR 0 (entry 164): stato pipeline ramo PdC. */
  stato_pipeline_pdc: StatoPipelinePdc;
  /** Sprint 8.0 MR 0 (entry 164): stato pipeline ramo Manutenzione. */
  stato_manutenzione: StatoManutenzione;
  km_max_giornaliero: number | null;
  km_max_ciclo: number | null;
  n_giornate_default: number;
  /** Sprint 7.8: lunghezza minima dei giri (soft, sotto solo per chiusure). */
  n_giornate_min: number;
  /** Sprint 7.8: lunghezza massima dei giri (hard cap). */
  n_giornate_max: number;
  fascia_oraria_tolerance_min: number;
  strict_options_json: StrictOptions;
  stazioni_sosta_extra_json: string[];
  created_by_user_id: number | null;
  /** Backend entry 88: popolato via JOIN con `app_user`, `null` se utente eliminato. */
  created_by_username: string | null;
  created_at: string;
  updated_at: string;
}

export interface FiltroRegolaPayload {
  campo: string;
  op: string;
  valore: unknown;
}

export interface ComposizioneItemPayload {
  materiale_tipo_codice: string;
  n_pezzi: number;
}

export interface ProgrammaRegolaAssegnazioneRead {
  id: number;
  programma_id: number;
  filtri_json: FiltroRegolaPayload[];
  composizione_json: ComposizioneItemPayload[];
  is_composizione_manuale: boolean;
  materiale_tipo_codice: string | null;
  numero_pezzi: number | null;
  priorita: number;
  /** Sprint 7.7 MR 1: cap km del ciclo specifico per regola (es. ETR526 ~4500). */
  km_max_ciclo: number | null;
  note: string | null;
  created_at: string;
}

export interface ProgrammaDettaglioRead extends ProgrammaMaterialeRead {
  regole: ProgrammaRegolaAssegnazioneRead[];
}

export interface ProgrammaRegolaAssegnazioneCreate {
  filtri_json?: FiltroRegolaPayload[];
  composizione: ComposizioneItemPayload[];
  is_composizione_manuale?: boolean;
  priorita?: number;
  /** Sprint 7.7 MR 1: cap km del ciclo specifico per regola (opzionale). */
  km_max_ciclo?: number | null;
  note?: string | null;
}

export interface ProgrammaMaterialeCreate {
  nome: string;
  valido_da: string;
  valido_a: string;
  km_max_giornaliero?: number | null;
  km_max_ciclo?: number | null;
  n_giornate_default?: number;
  /** Sprint 7.8: default 4. */
  n_giornate_min?: number;
  /** Sprint 7.8: default 12. */
  n_giornate_max?: number;
  fascia_oraria_tolerance_min?: number;
  strict_options_json?: Partial<StrictOptions>;
  stazioni_sosta_extra_json?: string[];
  regole?: ProgrammaRegolaAssegnazioneCreate[];
}

export interface ProgrammaMaterialeUpdate {
  nome?: string;
  valido_da?: string;
  valido_a?: string;
  km_max_giornaliero?: number | null;
  km_max_ciclo?: number | null;
  n_giornate_default?: number;
  /** Sprint 7.8: aggiorna lunghezza minima (soft) dei giri. */
  n_giornate_min?: number;
  /** Sprint 7.8: aggiorna lunghezza massima (hard) dei giri. */
  n_giornate_max?: number;
  fascia_oraria_tolerance_min?: number;
  strict_options_json?: Partial<StrictOptions>;
  stazioni_sosta_extra_json?: string[];
}

export interface ListProgrammiParams {
  stato?: ProgrammaStato;
}

function buildQuery(params: ListProgrammiParams): string {
  const search = new URLSearchParams();
  if (params.stato !== undefined) search.set("stato", params.stato);
  const qs = search.toString();
  return qs.length > 0 ? `?${qs}` : "";
}

export async function listProgrammi(
  params: ListProgrammiParams = {},
): Promise<ProgrammaMaterialeRead[]> {
  return apiJson<ProgrammaMaterialeRead[]>(`/api/programmi${buildQuery(params)}`, {
    method: "GET",
  });
}

export async function getProgramma(id: number): Promise<ProgrammaDettaglioRead> {
  return apiJson<ProgrammaDettaglioRead>(`/api/programmi/${id}`, { method: "GET" });
}

export async function createProgramma(
  payload: ProgrammaMaterialeCreate,
): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>("/api/programmi", {
    method: "POST",
    body: payload,
  });
}

export async function updateProgramma(
  id: number,
  payload: ProgrammaMaterialeUpdate,
): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}`, {
    method: "PATCH",
    body: payload,
  });
}

export async function pubblicaProgramma(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/pubblica`, {
    method: "POST",
  });
}

// =====================================================================
// Builder run (Sprint 7.9 MR 11C, entry 116)
// =====================================================================

export interface BuilderRunRead {
  id: number;
  programma_id: number;
  localita_codice: string;
  eseguito_at: string;
  eseguito_da_user_id: number | null;
  n_giri_creati: number;
  n_giri_chiusi: number;
  n_giri_non_chiusi: number;
  n_corse_processate: number;
  n_corse_residue: number;
  n_eventi_composizione: number;
  n_incompatibilita_materiale: number;
  warnings_json: string[];
  force: boolean;
}

export async function getLastBuilderRun(
  programmaId: number,
): Promise<BuilderRunRead | null> {
  return apiJson<BuilderRunRead | null>(
    `/api/programmi/${programmaId}/last-run`,
    { method: "GET" },
  );
}

export async function archiviaProgramma(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/archivia`, {
    method: "POST",
  });
}

export async function addRegola(
  programmaId: number,
  payload: ProgrammaRegolaAssegnazioneCreate,
): Promise<ProgrammaRegolaAssegnazioneRead> {
  return apiJson<ProgrammaRegolaAssegnazioneRead>(`/api/programmi/${programmaId}/regole`, {
    method: "POST",
    body: payload,
  });
}

export async function deleteRegola(programmaId: number, regolaId: number): Promise<void> {
  await apiJson<void>(`/api/programmi/${programmaId}/regole/${regolaId}`, {
    method: "DELETE",
  });
}

// =====================================================================
// Pipeline state machine — Sprint 8.0 MR 0 (entry 164)
// =====================================================================

export interface SbloccaProgrammaPayload {
  ramo: "pdc" | "manutenzione";
  motivo?: string | null;
}

export async function confermaMateriale(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/conferma-materiale`, {
    method: "POST",
  });
}

export async function confermaPdc(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/conferma-pdc`, {
    method: "POST",
  });
}

export async function confermaPersonale(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/conferma-personale`, {
    method: "POST",
  });
}

export async function pubblicaVistaPdc(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/pubblica-vista-pdc`, {
    method: "POST",
  });
}

export async function confermaManutenzione(id: number): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(
    `/api/programmi/${id}/conferma-manutenzione`,
    { method: "POST" },
  );
}

export async function sbloccaProgramma(
  id: number,
  payload: SbloccaProgrammaPayload,
): Promise<ProgrammaMaterialeRead> {
  return apiJson<ProgrammaMaterialeRead>(`/api/programmi/${id}/sblocca`, {
    method: "POST",
    body: payload,
  });
}

// =====================================================================
// Auto-assegna persone — Sub-MR 2.bis-a (Sprint 8.0)
// =====================================================================

/**
 * Body per `POST /api/programmi/{id}/auto-assegna-persone`.
 *
 * Entrambi i campi sono opzionali. Se omessi, il backend usa
 * `programma.valido_da` / `programma.valido_a`. Le date sono ISO
 * `YYYY-MM-DD`. Validatore Pydantic richiede `data_da ≤ data_a`.
 */
export interface AutoAssegnaPersonePayload {
  data_da?: string | null;
  data_a?: string | null;
}

/** Una nuova assegnazione persona → giornata creata dal greedy. */
export interface AssegnazioneCreata {
  persona_id: number;
  turno_pdc_giornata_id: number;
  data: string;
}

/**
 * Una giornata non coperta dall'algoritmo + motivo. Allineato con enum
 * server-side `MotivoMancanza` (vedi
 * `colazione.domain.normativa.assegnazione_persone`).
 */
export type MotivoMancanza =
  | "nessun_pdc_deposito"
  | "tutti_indisponibili"
  | "tutti_gia_assegnati"
  | "tutti_riposo_intraturno_violato"
  | "nessun_pdc_candidato";

export interface MancanzaAuto {
  turno_pdc_giornata_id: number;
  turno_pdc_id: number;
  data: string;
  motivo: MotivoMancanza;
}

/**
 * Tipo del warning soft. Allineato con enum server-side `TipoWarningSoft`.
 */
export type TipoWarningSoft =
  | "fr_cap_settimana_superato"
  | "fr_cap_28gg_superato"
  | "riposo_settimanale_violato"
  | "primo_giorno_post_riposo_mattina";

export interface WarningSoft {
  persona_id: number;
  data: string;
  tipo: TipoWarningSoft;
  descrizione: string;
}

/** Response dell'auto-assegna. KPI principale: `delta_copertura_pct`. */
export interface AutoAssegnaPersoneResponse {
  finestra_data_da: string;
  finestra_data_a: string;
  n_giornate_totali: number;
  n_giornate_coperte: number;
  n_assegnazioni_create: number;
  delta_copertura_pct: number;
  assegnazioni: AssegnazioneCreata[];
  mancanze: MancanzaAuto[];
  warning_soft: WarningSoft[];
}

export async function autoAssegnaPersone(
  id: number,
  payload: AutoAssegnaPersonePayload,
): Promise<AutoAssegnaPersoneResponse> {
  return apiJson<AutoAssegnaPersoneResponse>(
    `/api/programmi/${id}/auto-assegna-persone`,
    { method: "POST", body: payload },
  );
}

// =====================================================================
// Assegna manuale (override) — Sub-MR 2.bis-b (Sprint 8.0)
// =====================================================================

/**
 * Body per `POST /api/programmi/{id}/assegna-manuale`.
 *
 * Override consapevole del pianificatore: bypassa i vincoli HARD del
 * greedy (sede, indisp, riposo intraturno) ma rispetta uniqueness
 * (no doppia persona/data, no doppia giornata/data).
 */
export interface AssegnaManualePayload {
  persona_id: number;
  turno_pdc_giornata_id: number;
  /** ISO YYYY-MM-DD. */
  data: string;
}

export async function assegnaManuale(
  id: number,
  payload: AssegnaManualePayload,
): Promise<AssegnazioneCreata> {
  return apiJson<AssegnazioneCreata>(
    `/api/programmi/${id}/assegna-manuale`,
    { method: "POST", body: payload },
  );
}

File: backend/src/colazione/models/programmi.py
"""Strato 1bis — programma materiale (input umano del pianificatore).

Definito in `docs/PROGRAMMA-MATERIALE.md` v0.2 (Sprint 4.0). Il
programma materiale è il registro autorevole delle scelte di
programmazione del pianificatore: per ogni gruppo di corse PdE
(filtrate per linea/direttrice/categoria/orario/giorno_tipo/...),
quale rotabile usare e in quante quantità.

Multi-tenant: ogni `azienda` può avere più programmi (per periodo,
linea, ecc.). Solo i programmi `attivo` sono usati dall'algoritmo
A (PdE → Giro Materiale).

Sprint 7.3: il campo `stagione` è stato rimosso. Il filtro temporale
delle corse è data-driven via `valido_in_date_json` di ogni corsa
(verità del PdE) intersecato con `[valido_da, valido_a]` del programma.
Le 3 stagioni (`invernale/estiva/agosto`) restano solo a livello di
`corsa_composizione.stagione` (composizione-tipo del materiale).
"""

from datetime import date, datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.exc import MissingGreenlet
from sqlalchemy.orm import Mapped, mapped_column, relationship

from colazione.db import Base

if TYPE_CHECKING:
    from colazione.models.auth import AppUser


class ProgrammaMateriale(Base):
    """Intestazione del programma materiale per una azienda + finestra.

    Container di N `ProgrammaRegolaAssegnazione`. Lo stato `bozza` →
    editing, `attivo` → usato dall'algoritmo, `archiviato` → solo storico.
    """

    __tablename__ = "programma_materiale"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    azienda_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("azienda.id", ondelete="RESTRICT")
    )

    nome: Mapped[str] = mapped_column(Text)
    valido_da: Mapped[date] = mapped_column(Date)
    valido_a: Mapped[date] = mapped_column(Date)
    stato: Mapped[str] = mapped_column(String(20), default="bozza")

    # Sprint 8.0 MR 0 (entry 164, migration 0032): pipeline state machine
    # per la concatenazione tra ruoli. Validazione applicativa via enum
    # ``StatoPipelinePdc`` in ``colazione.domain.pipeline``; lato DB
    # protetto da CHECK constraint ``programma_materiale_stato_pipeline_pdc_check``.
    stato_pipeline_pdc: Mapped[str] = mapped_column(
        String(40),
        default="PDE_IN_LAVORAZIONE",
        server_default="PDE_IN_LAVORAZIONE",
    )
    # Ramo manutenzione, parallelo e indipendente dal ramo PdC. Si attiva
    # quando ``stato_pipeline_pdc >= MATERIALE_CONFERMATO`` ma non blocca
    # le transizioni del ramo principale.
    stato_manutenzione: Mapped[str] = mapped_column(
        String(40), default="IN_ATTESA", server_default="IN_ATTESA"
    )

    # Parametri globali
    km_max_giornaliero: Mapped[int | None] = mapped_column(Integer)
    # Cap km cumulati sul ciclo intero (NON giornaliero). Quando raggiunto,
    # il giro chiude con rientro programmato. Sprint 5.1 (migration 0007).
    # Configurato dal pianificatore per ogni programma; tipici 5000-10000.
    km_max_ciclo: Mapped[int | None] = mapped_column(Integer)
    n_giornate_default: Mapped[int] = mapped_column(Integer, default=1)
    # Sprint 7.8 (migration 0019): range lunghezza giri generati dal
    # builder. `n_giornate_min` è SOFT (il builder scende sotto solo
    # per giri di "chiusura" delle corse residue a fine pool).
    # `n_giornate_max` è HARD (nessun giro supera questo numero).
    # Decisione utente 2026-05-03: default min=4, max=12.
    n_giornate_min: Mapped[int] = mapped_column(Integer, default=4, server_default="4")
    n_giornate_max: Mapped[int] = mapped_column(Integer, default=12, server_default="12")
    fascia_oraria_tolerance_min: Mapped[int] = mapped_column(Integer, default=30)

    # Strict mode granulare (vedi PROGRAMMA-MATERIALE.md §2.7)
    strict_options_json: Mapped[dict[str, Any]] = mapped_column(JSONB, default=dict)

    # Sprint 5.6 (migration 0008): codici stazione PdE che il pianificatore
    # ammette come SOSTA NOTTURNA del materiale di questo programma, oltre
    # alla whitelist sede e ai depot PdC. La lista finale del builder è
    # whitelist sede ∪ depot PdC ∪ stazioni_sosta_extra. Default `[]`.
    stazioni_sosta_extra_json: Mapped[list[str]] = mapped_column(
        JSONB, default=list, server_default="[]"
    )

    # Tracking
    created_by_user_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("app_user.id"))
    created_by: Mapped["AppUser | None"] = relationship(foreign_keys=[created_by_user_id])
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    @property
    def created_by_username(self) -> str | None:
        """Username del creatore se la relazione è eager-loaded.

        Restituisce ``None`` se la relazione non è stata caricata
        (per evitare lazy-load implicito in contesto async). Le query
        che vogliono questo campo devono usare `joinedload(...)`.
        Entry 88 (TN-UPDATE).
        """
        try:
            return self.created_by.username if self.created_by is not None else None
        except MissingGreenlet:
            return None


class ProgrammaRegolaAssegnazione(Base):
    """Singola regola di assegnazione: filtri AND → (tipo, n_pezzi).

    Una regola matcha una corsa se TUTTI i filtri in `filtri_json`
    sono soddisfatti. Vedi `risolvi_corsa()` in `domain/builder_giro/`
    per l'algoritmo (Sprint 4.2).

    Schema `filtri_json`:
        [
          {"campo": "codice_linea", "op": "eq", "valore": "S5"},
          {"campo": "fascia_oraria", "op": "between", "valore": ["04:00", "15:59"]},
          {"campo": "giorno_tipo", "op": "in", "valore": ["feriale"]}
        ]

    Validazione applicativa via Pydantic in `schemas/programmi.py`.
    """

    __tablename__ = "programma_regola_assegnazione"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    programma_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("programma_materiale.id", ondelete="CASCADE")
    )

    filtri_json: Mapped[list[Any]] = mapped_column(JSONB, default=list)

    # Composizione del rotabile: lista
    # `[{materiale_tipo_codice, n_pezzi}, ...]` (Sprint 5.1, migration
    # 0007). Sostituisce a regime i campi legacy `materiale_tipo_codice`
    # + `numero_pezzi`. Lista di 1 elemento per regola single-material;
    # 2+ elementi per composizione mista (es. ETR526+ETR425). Validata
    # da `ComposizioneItem` in schemas/programmi.py.
    composizione_json: Mapped[list[Any]] = mapped_column(JSONB, default=list)
    # Override manuale: se True, bypass del check
    # `materiale_accoppiamento_ammesso` per composizioni custom decise
    # dal pianificatore (Sprint 5.1).
    is_composizione_manuale: Mapped[bool] = mapped_column(Boolean, default=False)

    # Campi LEGACY (deprecati ma non rimossi). Resi nullable in 0007.
    # Letti ancora da `risolvi_corsa()` fino a Sprint 5.5; sulle nuove
    # righe popolati dal primo elemento di `composizione_json` (handler
    # API) per retrocompat.
    materiale_tipo_codice: Mapped[str | None] = mapped_column(
        String(50), ForeignKey("materiale_tipo.codice", ondelete="RESTRICT")
    )
    numero_pezzi: Mapped[int | None] = mapped_column(Integer)

    priorita: Mapped[int] = mapped_column(Integer, default=60)
    # Sprint 7.7 MR 1 (migration 0014): cap km del ciclo specifico per
    # questa regola/materiale. Il builder usa
    # `regola.km_max_ciclo OR programma.km_max_ciclo OR DEFAULT_*` come
    # cap effettivo del giro generato da questa regola. Ogni materiale
    # ha autonomie diverse (ETR526 ~4500, E464 ~6000, ATR803 ~ecc).
    km_max_ciclo: Mapped[int | None] = mapped_column(Integer)
    note: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class BuilderRun(Base):
    """Esito persistito di una run del builder (Sprint 7.9 MR 11C, entry 116).

    Ogni invocazione di ``genera_giri`` produce una riga in questa
    tabella che conserva stats + warnings, così il pianificatore può:
    - Vedere perché il run ha prodotto 0 giri (warnings espliciti tipo
      "sede non coerente con la regola").
    - Confrontare run successivi (regole modificate, sede cambiata).
    - Esporre la copertura PdE (corse_processate / corse_residue).

    Decisione utente 2026-05-04: warning del builder oggi sono persi
    dopo la response. La persistenza permette di mostrare dashboard
    "Ultimo run" + storico run senza re-eseguire.
    """

    __tablename__ = "builder_run"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    programma_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("programma_materiale.id", ondelete="CASCADE"),
        nullable=False,
    )
    azienda_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("azienda.id", ondelete="CASCADE"), nullable=False
    )
    localita_codice: Mapped[str] = mapped_column(String(64), nullable=False)
    eseguito_da_user_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("app_user.id", ondelete="SET NULL"), nullable=True
    )
    eseguito_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    n_giri_creati: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    n_giri_chiusi: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    n_giri_non_chiusi: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    n_corse_processate: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    n_corse_residue: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    n_eventi_composizione: Mapped[int] = mapped_column(
        Integer, default=0, nullable=False
    )
    n_incompatibilita_materiale: Mapped[int] = mapped_column(
        Integer, default=0, nullable=False
    )
    warnings_json: Mapped[list[str]] = mapped_column(JSONB, default=list, nullable=False)
    force: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
